File: week 2/fibonacci_huge/fibonacci_huge.py
# Uses python3

def get_fibonaccihuge(n, m):
    # write your code here
    if (n == 0):
        return 0
		
    # fib = [0, 1]
    # mod = [0, 1]
    fib = {0:0, 1:1}
    mod = {0:0, 1:1}
	
    i = 2
    while True:
        # fib.insert(i, mod[i-1] + mod[i-2])
        # mod.insert(i, fib[i] % m)
        fib.setdefault(i, mod.get(i-1) + mod.get(i-2))
        mod.setdefault(i, fib.get(i) % m)
        if(mod[i-1] == 0 and mod[i] == 1):
            break
        i = i+1
    remainder = n % (i-1) # The period length is i-1
    # return fib[remainder] % m
    return fib.get(remainder) % m

if __name__ == '__main__':
    n, m = map(int, raw_input().split())
    print(get_fibonaccihuge(n, m))

File: week 2/fibonacci_huge/fibonacci_huge.cpp
#include <iostream>
#include <vector>

using std::vector;

long long get_fibonaccihuge(long long n, long long m) {
  //write your code here
	if(n == 0)
		return 0;
  
	vector<long> fib(n+1);
	vector<long> mod(m+1);
	
	fib[0] = 0;
	fib[1] = 1;
	mod[0] = 0;
	mod[1] = 1;

	int i=2;
	while(true) {
		fib[i] = mod[i-1] + mod[i-2];
		mod[i] = fib[i] % m;
		if(mod[i-1] == 0 && mod[i] == 1)
			break;
		i++;
	}
	int remainder = (int)(n % (i-1)); // The period length is i-1
	return fib[remainder] % m;
}

int main() {
    long long n, m;
    std::cin >> n >> m;
    std::cout << get_fibonaccihuge(n, m) << '\n';
}
